n_neighbors defaults for Manifold Visualizer

jimmyshah opened the pull request:
This PR fixes the n_neighbors concern from Issue #437 , so that the default is None, and the parameter is required for all transformers except for mds and tsne

lwgray commented:
@jimmyshah Thank you for the PR.  As we work our way through our backlog of Issues and PRs... I promise that we will get to this PR asap.

lwgray commented:
@jimmyshah as you merge each of these can you check them off the list in #437 and include the PR number that was merged

jimmyshah commented:
@lwgray Apologies for the delay! How do I check off the items in the list? Also, I think this build is failing, so I might need to make some changes? Do you any advice that might help me better understand the issue?

lwgray commented:
@jimmyshah  Sorry but disregard the checklist comment(it is something that only I can do). I think Maybe just going through the PR process will help you get a  better understanding.

jimmyshah commented:
@lwgray No worries at all, thank you for the assistance!

lwgray commented:
No worries, We’re here to help 😸

jimmyshah commented:
@lwgray I'm having some trouble understanding the test failures. Do you have any advice on what I can do differently?

lwgray commented:
@jimmyshah  @bbengfort will be reviewing this and will maybe have some better insight into the errors.  However, I am not sure when @bbengfort will be able to get to this but he will as soon as he can.

jimmyshah commented:
@bbengfort Thank you so much for the detailed response! This is incredibly helpful, and I'll see if I can tackle the issues later tonight. I'm really excited to contribute and I appreciate the help.

jimmyshah commented:
@bbengfort  I spent some time digging into n_neighbors, and I found that the parameter is dependent on sklearn.manifold. Of all of the manifold transformers, tsne, mds, and spectral do not use the n_neighbors parameter. I tried creating an if statement and raising a YellowBrickValueWarning, but this was still causing a few tests to fail. Would an alternative option be to simply print out a warning message? Please let me know your thoughts and if there is further exploration needed.

bbengfort commented on yellowbrick/features/manifold.py:
```diff
@@ -236,10 +236,16 @@ def manifold(self, transformer):
 
+            if (transformer not in ('tsne','mds','spectral') and self.n_neighbors is None):
+                self.n_neighbors = 10
+                print(
```
Rather than doing a `print`, I'd prefer to issue a warning:

```python
import warnings 

from yellowbrick.exceptions import YellowbrickWarning

warnmsg = "using n_neighbors={}; please explicitly pass this argument for the '{}' manifold".format(self.n_neighbors, str(transformer))
warnings.warn(warnmsg, YellowbrickWarning)
```

We will then also need to add a test to make sure this warning is issued.

jimmyshah commented:
@bbengfort  Good catch! The spectral transformer does have the n_neighbors parameter, I have removed it in the latest push.

For the other manifolds, lle. ltsa, hessian, isomap, and modified have 5 for their default n_neighbors values, according the scikit-learn documentation. I've updated this accordingly as well in the docstring. However, hessian requires n_neighbors > n_components * (1 + (n_components + 1) / 2 . 

I'm having some trouble understanding the solution and writing the correct tests...

jimmyshah commented:
Are there any updates for this? Any suggestions or guidance would be sincerely appreciated

bbengfort commented on yellowbrick/features/manifold.py:
```diff
@@ -218,6 +220,7 @@ def __init__(
         self.alpha = alpha
         self.random_state = random_state
         self.manifold = manifold # must be set last
+        self.n_components = None
```
`n_components` must be `2` (eventually we could make it `3`) ... this is how we set the dimensionality to draw the figure on a 2D graph. Therefore we cannot set this property, making it so the user could change the dimensionality -- we would no longer be able to draw the figure!

bbengfort commented on yellowbrick/features/manifold.py:
```diff
@@ -236,16 +239,21 @@ def manifold(self, transformer):
+            if (transformer not in ('tsne','mds','hessian') and self.n_neighbors is None):
+                self.n_neighbors = 5
+                warnmsg = "using n_neighbors={}; please explicitly pass this argument for the '{}' manifold".format(self.n_neighbors, str(transformer))
+                warnings.warn(warnmsg, YellowbrickWarning)
```
Thank you for adding the warning here! However, I think we should be more explicit, e.g.

```python
# 2 components is required for 2D plots, could optionally make it three for 3D plots in future
n_components = 2

requires_default_neighbors = {'lle', 'ltsa', 'isomap', 'hessian', 'spectral', 'modified'}
if self.n_neighbors is None and transformer in requires_default_neighbors:
    if transformer == 'hessian':
        self.n_neighbors = 1 + (n_components * (1 + (n_components + 1) / 2))
    else:
        self.n_neighbors = 5 

    warnmsg = "using n_neighbors={}; please explicity specify for the '{}' manifold".format(self.n_neighbors, transfomrer)
    warnings.warn(warnmsg, YellowbrickWarning)


```

bbengfort commented on yellowbrick/features/manifold.py:
```diff
@@ -236,16 +239,21 @@ def manifold(self, transformer):
-                "n_components": 2,
-                "n_neighbors": self.n_neighbors,
+                "n_components": self.n_components,
+                "n_neighbors": 1 + (self.n_components * (1 + (self.n_components + 1) / 2)),
```
This would become

```
params = {
    "n_components": n_components,
    "n_neighbors": self.n_neighbors,
}
```

bbengfort commented:
@jimmyshah Thanks again for all your hard work on this issue!

jimmyshah commented:
@bbengfort Thank you so much for the detailed response! No problem at all, I completely understand and appreciate your help. I think I have a much better picture of the issue, and will spend some time tomorrow/this weekend tackling it.

bbengfort commented:
@jimmyshah great news! And please ping me again when you're ready -- thank you!

jimmyshah commented:
@bbengfort Thank you again for all of your help, I really do appreciate it! I made the changes you suggested, but have two tests failing. One of the tests seems to check the type of n_neighbors, and the other is an image comparison. I'm unsure of how to resolve these issues. Do the tests need to be changed or does my code need to be modified?

rebeccabilbro commented:
Thank you so much @jimmyshah for this second contribution and @bbengfort for helping get this across the finish line! I went ahead and merged this in now that the tests are passing 🎉 hope that's ok!

 @jimmyshah &mdash; regarding your question about the tests; yes, generally anytime an update is made to the code for a visualizer, you should expect to either (1) create a new unit test (2) update the existing unit test or (3) in the case of a visual comparison test, update the baseline image for the existing test. While this can seem a bit tricky at first, if you're thinking of contributing again in the future, we'd be very grateful if in addition to reviewing @bbengfort's updates to the tests here in this PR, you could take a deep dive into our contributor documentation, including:

 - [how to run the tests locally](https://www.scikit-yb.org/en/develop/contributing/developing_visualizers.html#running-the-test-suite) on your own machine before pushing to a PR branch
 - [how our visual comparison tests work](https://www.scikit-yb.org/en/develop/contributing/developing_visualizers.html#image-comparison-tests), including how to update baseline images
 - [what our code conventions are](https://www.scikit-yb.org/en/develop/contributing/advanced_development_topics.html#code-conventions) including things like how to handle `random_state` 

It might even be a good idea to read through some of other recently merged PRs! We're of course happy to help newcomers, but since we are all volunteers doing this in our spare time 😅, we're unfortunately spread a little thin and not always able to respond quickly &mdash; for that reason we definitely appreciate when veteran contributors can dive a bit deeper into our API and documentation and deepen their intuition about how to go about solving YB problems!

jimmyshah commented:
@rebeccabilbro my apologies, just getting to this now! Thank you for the help, and I can definitely look into the contributor documentation. I'm fairly new to contributing to open source libraries, so I apologize for any inconveniences. Thanks again!
